### PR TITLE
feat(identity): add `From` impls for specific keypair types

### LINKED ISSUE #3618: Allow converting ed25519::Keypair to enum Keypair
## Description

Add a From impl for libp2p::core::identity::Keypair for struct ed25519::Keypair.

## Motivation

This is from our use case in [forest](https://github.com/ChainSafe/forest/blob/6b8a3cf0db4313ea1efb4f4577622234ee91e5fb/forest/daemon/src/daemon.rs#L90), we require converting a `ed25519::Keypair` struct to the enum `libp2p::core::identity::Keypair`. It would be helpful to have `From` impl for Keypair.

## Requirements

1.
2.
3.

## Are you planning to do it yourself in a pull request?

Yes

### PR DESCRIPTION
**Pull request by creativcoder**
## Description

Closes #3618.

## Change checklist

- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: identity/src/ecdsa.rs
Comment (thomaseizinger) [inline identity/src/ecdsa.rs]: Having this in here creates a cyclic dependency between the two modules. I'd rather have the `From` impls next to the enum.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
One request, otherwise LGTM.

Can you add a changelog entry please and bump the patch version?

File: identity/src/ecdsa.rs
Comment (creativcoder) [reply identity/src/ecdsa.rs]: Moved to keypair.rs.

**Comment by creativcoder** [comment]
Sure, `changelog.md` at core/ or should i create a new one under identity/ ?

File: identity/src/keypair.rs
Comment (thomaseizinger) [inline identity/src/keypair.rs]:
```suggestion
#[cfg(feature = "ecdsa")]
impl From<ecdsa::Keypair> for Keypair {
    fn from(kp: ecdsa::Keypair) -> Self {
        #[allow(deprecated)]
        Keypair::Ecdsa(kp)
    }
}
```

I believe we can make this a bit more concise, same applies to the others.

**Comment by thomaseizinger** [comment]
> Sure, `changelog.md` at core/ or should i create a new one under identity/ ?

New one in `identity` please. Add a link the root changelog to it too please!

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Slight adjustment, otherwise LGTM

File: identity/src/keypair.rs
Comment (creativcoder) [reply identity/src/keypair.rs]: fixed

File: identity/CHANGELOG.md
Comment (thomaseizinger) [inline identity/CHANGELOG.md]:
```suggestion
# 0.1.1 [unreleased]
```

File: identity/CHANGELOG.md
Comment (thomaseizinger) [inline identity/CHANGELOG.md]:
```suggestion
- Add `From` impl for specific keypairs.
  See [PR 3626].

[PR 3626]: https://github.com/libp2p/rust-libp2p/pull/3626
```

File: identity/src/keypair.rs
Comment (thomaseizinger) [inline identity/src/keypair.rs]:
```suggestion
#[cfg(all(feature "rsa", not(target_arch = "wasm32")))]
```

rsa needs some more specific flags.

Comment (creativcoder) [reply identity/src/keypair.rs]: added

File: identity/CHANGELOG.md
Comment (creativcoder) [reply identity/CHANGELOG.md]: fixed

File: identity/CHANGELOG.md
Comment (creativcoder) [reply identity/CHANGELOG.md]: fixed

**Review by thomaseizinger — APPROVED** [review]
Nice! Thank you for your work and patience :)

**Comment by thomaseizinger** [comment]
@mxinden Do you mind if I have a go at releasing this as soon as it is merged?

File: identity/src/keypair.rs
Comment (mxinden) [inline identity/src/keypair.rs:271]: How do we implement the `From` once we removed the deprecated `Keypair::Ecdsa` @thomaseizinger?

**Review by mxinden — COMMENTED** [review]
No objections to releasing, though I have a question below.

For the release process, I suggest removing the `[unreleased]`, then merging `master` into `v0.51` (fast forward, no merge commit needed), and then releasing from there.

File: identity/src/keypair.rs
Comment (thomaseizinger) [reply identity/src/keypair.rs:271]: What do you mean? ECDSA is not deprecated, only the direct access to the enum variant is.

Once the type is opaque, the enum will still exist, just not publicly accessible and so we can still implement the `From` trait.

Comment (mxinden) [reply identity/src/keypair.rs:271]: Makes sense. Thanks. Feel free to go ahead with the release @thomaseizinger.

**Comment by thomaseizinger** [comment]
> For the release process, I suggest removing the `[unreleased]`, then merging `master` into `v0.51` (fast forward, no merge commit needed), and then releasing from there.

Sounds good. I started with a prepare PR here: https://github.com/libp2p/rust-libp2p/pull/3626#pullrequestreview-1344657608

Once merged, I'll locally fast-forward merge master into `v0.51`, do a release and push the resulting commits and tags.

**Comment by thomaseizinger** [comment]
Unfortunately, I couldn't finish the release because `libp2p-identity` only has you as an owner @mxinden: https://crates.io/crates/libp2p-identity

**Comment by thomaseizinger** [comment]
The PR is merged and `v0.51` is updated and pushed, all you'd need to do is run the release command please.

**Comment by mxinden** [comment]
> Unfortunately, I couldn't finish the release because `libp2p-identity` only has you as an owner @mxinden: https://crates.io/crates/libp2p-identity

My bad. Added the rust-libp2p-maintainers group just now.

> The PR is merged and `v0.51` is updated and pushed, all you'd need to do is run the release command please.

I will get to it next week. @thomaseizinger feel free to race me to it in case you have time beforehand.

**Comment by thomaseizinger** [comment]
[Tagged](https://github.com/libp2p/rust-libp2p/releases/tag/libp2p-identity-v0.1.1) and [published](https://crates.io/crates/libp2p-identity/0.1.1).

**Comment by mxinden** [comment]
:tada: first release by @thomaseizinger :tada:

**Pull request merged**
